### ISSUE TITLE
Weather station refuses to download files

### ISSUE THREAD

**Issue by phanus** [cetrespeo/ESP32-E-Paper-Weather#4]
Hello Cetrespeo! I appreciate all of your hard work on the weatherstation. I used revision 2.31R and it worked for several weeks really well. Today I decided to update to the latest version 2.34. After the update, the weather station is unable to download any file. See attched log. I tried to downgrade back to 2.31R but it doesn't work either. Internet connection is working and the weather station connects to my network just fine - it is able to sync the time over NTP just fine, but it is not able to download files fro amazon cloud nor weather forecast. If I paste the forecast link to the browser, it works OK. Please advice what could be wrong, I'm using lolin32, latest libraries except the arduino json and firebase, for those two I use recommended versions. 
Thank you in advance,
Cheers!
Petr H.

Listing directory: /
  FILE: /wifi.txt	SIZE: 58B 
  FILE: /jVars.txt	SIZE: 408B 
  FILE: /resets.txt	SIZE: 1B 
  FILE: /jLog.txt	SIZE: 179B 
  FILE: /jVtg.txt	SIZE: 1600B 
  Total/Used/Free:   1816/4/1812 kB
--------------------------------------------
{R:'/jVars.txt'>408B}~LVars_SPFS 408B~{R:'/jVtg.txt'>1600B}~LVtg_SPFS 1600B~{R:'/jLog.txt'>179B}~LLog_SPFS 179B~
Loading Wifi.txt...{R:'/wifi.txt'>58B}  Connecting to WIFI,mynetworknamecensored.. (pwd:31 chrs long)-> @192.168.1.111 RSSI:-68dBm   NTP . 2021/03/01_21:45:48 Ok.

Copying from data/edit.htm to /edit.htm
Connecting to: s3.eu-west-3.amazonaws.com
0 bytes downloaded.
 {D'/edit.htm'}

Copying from data/ace.js to /ace.js
Connecting to: s3.eu-west-3.amazonaws.com
0 bytes downloaded.
 {D'/ace.js'}

Copying from data/mode-html.js to /mode-html.js
Connecting to: s3.eu-west-3.amazonaws.com
0 bytes downloaded.
 {D'/mode-html.js'}
  SPIFFS: 1 WifiSsids, Vtg{0>0>2366>10000}  Wifi:1,Firebase:0,Dev:0 
  Updated 1 times, on core 1 since never.{R:'/resets.txt'>1B}**MENU:1 reboots-> Standard  {W'/resets.txt'>1B} [FST mVtg:2363@53ms],  [New VtgMax = 2363] {$3 : 0} {W'/resets.txt'>1B}Entering Bootmode = Standard
 GetWF:214kB free, FB_GetWeatherJson , getting json_Time failed :  No FB WJ!,  <Download openweathermap>   Connecting to {api.openweathermap.org} , 
 ERROR **Connection failed for JSON {http://api.openweathermap.org/data/2.5/onecall?lat=49.55&lon=16.57&appid=keycensored&lang=en&units=metric&exclude=minutely,daily} **
  Connecting to {api.openweathermap.org} , 
 ERROR **Connection failed for JSON {http://api.openweathermap.org/data/2.5/onecall?lat=49.55&lon=16.57&appid=keycensored&lang=en&units=metric&exclude=minutely,hourly} **
{R:'/weather.txt' failed!}
 FB_GetWeatherJson , getting json_Time failed : , No weather data.#FB_OLD#
NO forecast, WF:1,FB:0,DV:0 _

[-> To sleep... 76 mins. @93 secs/74 mean , 24.19ºC PON>SLP74@93/74s/c {W'/jLog.txt'>179B} {W'/jVtg.txt'>1600B}]. Zzz
-------------------------------- -

**Comment by phanus** [comment]
Downgrading ESP32 boards package to ver 1.0.4 resolved the issue. Version 1.0.5 does not work.

**Comment by cetrespeo** [comment]
Hi phanus,

It seems weather data json cannot be downloaded from OWM. Please note the ERROR in your log
`ERROR **Connection failed for JSON {http://api.openweathermap.org/data/2.5/onecall?lat=49.55&lon=16.57&appid=keycensored&lang=en&units=metric&exclude=minutely,daily} **
`
I also see that the internal webpage files from AWS give `0 bytes downloaded`. The device seems to be unable to download anything.

Maybe espressif changed/broke something on 1.0.5 wifi connection. I will check.

Thanks for the feedback

EDIT: just checked ESP32 boards package 1.0,5, and application is also working fine as in 1.0.4.

**Comment by cetrespeo** [comment]
Hi phanus,

Also, can you please give the usual 3 letter abreviations of Czech day names?

`{"neděle", "pondělí", "úterý", "středa", "čtvrtek", "pátek", "sobota"}`

.. so I can implement Czech in the code. :-) It is nice to have feedback from people using this thingy :-)

Thanks

**Comment by phanus** [comment]
Hello cetrespeo, we usually use only 2 letter abbreviations for days. {"Ne", "Po", "Út", "St", "Čt", "Pá", "So"} and 3 letter abbreviations for months {"Led", "Úno", "Bře", "Dub", "Kvě", "Čer", "Čec", "Srp", "Zář", "Říj", "Lis", "Pro"}. Thanks

**Comment by phanus** [comment]
With the 1.0.5 ESP32 boards package I enabled verbose debug level and the WifiSlientSecure always return -1 when calling .connect method. Not sure why. ESP32 boards 1.0.4 works just fine.

**Comment by phanus** [comment]
I'm also curious - I have the wether station running on two 18650 cells in parallel, I have the analog pin connected via resistor divider to battery + terminal. The weatherstation shows battery icon slowly decreasing over the time, but it does not show estimated remaining days. At the very beginning it showed "FULL" but now it just shows "-". Will the remaining days start working over the time after collecting enough battery data? I tried to follow the battery monitoring code, but I don't fully understand it....

**Comment by phanus** [comment]
I tried to add the Czech names into the code and the weather station font does not support Czech accented characters which I suppose will be problem with openweathermap current conditions wording. For example "Rain" is in Czech "Déšť" but the accented characters are not displayed, so the display says: "Dt". :( I'm not sure if there is any font for U8G2 supporting Czech accented characters (full UTF8, maybe ISO-8859-1 or ISO-8859-2) or in general how to display accented characters.

**Comment by cetrespeo** [comment]
> I'm also curious - I have the wether station running on two 18650 cells in parallel, I have the analog pin connected via resistor divider to battery + terminal. The weatherstation shows battery icon slowly decreasing over the time, but it does not show estimated remaining days. At the very beginning it showed "FULL" but now it just shows "-". Will the remaining days start working over the time after collecting enough battery data? I tried to follow the battery monitoring code, but I don't fully understand it....

Hi, 

Sorry for not including a detailed explanation in the readme. Long story short. You need a full charge and full discharge cycle in order to get the battery graph correctly calculated and the estimated remaining days accurate.

_VtgMax & VtgMin (graph bar)_
The application needs a full charge and discharge cycle to know the boundaries of the battery. Batt level is measured with the actual value and the **VtgMax** and **VtgMin** values (the limits). It is dynamically done as I have used LiFePo, Li-Ion, NiMh, and all sort of batteries with this thing, and I needed it to be dynamic and adaptable to anything. Some reach up to 4,20V, some 4,25 and others just 4,10... Minimum values are funnier as some reach down to stable 3,4V and others die at 3,7V... 

To define **VtgMax**. first, when the device detects a charge (voltage increases and slope of variation is possitive) it waits until the charging slope is more or less 0 (no big increments). When the slope is near 0, it supposes maximum voltage and writes it down in the Vtg json (only if Vtg is > the already written VtgMax).

Then it establishes a default estimated min Voltage value (just estimative and really high) so it can make an initial idea of what is the battery level remaining.

When fully charged all li-ion batteries decrease really fast on the first mV. Then they have a rather slow slope down. When near to discharge that negative slope gets bigger indicating that your battery is near depletion.

So, I am looking continuously to the slope of the voltage. Whenever I see a rapid decrease, I update the **VtgMin** (only if is less than the already marked VtgMin), When slope gets too negative, I bypass the update as the values are not representative because I know the battery is about to die.

_Remaining days_
The remaining days are calculated with PeriodDrop and PeriodMax. PeriodDrop is the expected Voltage difference in a standard battery cycle (it is the valid range of battery life in V). PeriodMax is the time period in which this expected voltage drop is expected to happen (that is the expected time). 

I make the expected days not depending on the Voltage as the Voltage drop curve of a Li-ion is far from linear and it gave me not usable readings. So I just use voltage to define the valid range of voltage life, and then I measure the period in which that voltage drop happens. It seems messy, but is the best way to have stable predictions.

_Method_
Whenever your Vtg is near VtgMax and slope is near 0, it establishes VtgChg and tLastChg (voltage of the charge and time of last charge). Then, each display update cycle, the app compares the difference between actual time and time last charge (tNow-tLastChg) and compares it with the PeriodMax value. If (tNow-tLastChg)>PeriodMax (in hours) then it updates both (PeriodMax and PeriodDrop). Update is skipped if Voltage is out of stable curve (slope > certain levels) as Vtg values decrease really fast on it final moments and will make the PeriodDrop value not representative.

Whenever you need to calculate the remaining days of batt, you just make the difference between the tNow-tLastChg and the expected PeriodMax(in seconds), with a proportial correction to the VtgChg and VtgMax (to control non full charges). 

Hope that gives you a better idea. I am open to improvement proposals! :-)

Thanks mate

**Comment by cetrespeo** [comment]
> I tried to add the Czech names into the code and the weather station font does not support Czech accented characters which I suppose will be problem with openweathermap current conditions wording. For example "Rain" is in Czech "Déšť" but the accented characters are not displayed, so the display says: "Dt". :( I'm not sure if there is any font for U8G2 supporting Czech accented characters (full UTF8, maybe ISO-8859-1 or ISO-8859-2) or in general how to display accented characters.

I have similar problem here with accents. I agree it is a problem with the chosen font (chosen for its size). Maybe other fonts are compatible with full UTF8 but I didn't work it out. 

Please let me know if you find a compatible small font.

**Comment by phanus** [comment]
I was looking for UTF8 or Latin font in the U8G2 library and there are not many of those, but I eventually got it working with Helvetica _te font and generic _me font. Now it displays correctly "Oblačno" ("Clouds") and "Bře" ("Mar").

Here is the code for WS4c display:
Fonts
//GOODDISPLAY 4 BW//////////////////
#if defined(WS4) || defined(WS4k)
#include <GxGDEW042T2/GxGDEW042T2.h>
#endif
#if defined(WS4) || defined(WS4c)
#define DisableClock
const bool bWS75 = false, bWS58 = false, bWS42 = true, bWS29 = false;
const uint8_t* fU8g2_XS = u8g2_font_6x13_me;
const uint8_t* fU8g2_S = u8g2_font_helvB12_te;
const uint8_t* fU8g2_M = u8g2_font_helvB14_te;
const uint8_t* fU8g2_L = u8g2_font_helvB18_te;
const uint8_t* fU8g2_XL = u8g2_font_helvB24_te;
const uint8_t* fU8g2_XXL = u8g2_font_logisoso54_tf;
#endif

Months
  if (sLang == (String)("cz")) {
    const char* MonthNames[12] = {"Led", "Úno", "Bře", "Dub", "Kvě", "Čer", "Čec", "Srp", "Zář", "Říj", "Lis", "Pro"};
    return (MonthNames[month(local) - 1]);
  }

Days
  if (sLang == (String)("cz")) {
    const char* WeekDayNames[7] = {"Ne", "Po", "Út", "St", "Čt", "Pá", "So"};
    return WeekDayNames[iDay];
  }

**Comment by phanus** [comment]
...and I replaced 
const uint8_t* fU8g2_XXL = u8g2_font_logisoso54_tf; 
with 
const uint8_t* fU8g2_XXL = u8g2_font_fub42_tf;
It looks better next to helvetica.

**Comment by cetrespeo** [comment]
:-) nice!. Bold and clean. The main problem with these large fonts are their size. I am trying to maintain the app below 1230kB limit of the default app size.

Anyway I added to the source as option. Thanks Petr
